PatternSyntaxException on @Path definition with regexp

## Describe the bug
`@Path("/{pin: [A-Z][A-Z][A-Z0-9]{9}[0-9]}")` is not parsed correctly

### Expected behavior
regexp should work

### Actual behavior
```
2021-05-11 12:48:54,048 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (vert.x-worker-thread-1) Failed to start quarkus: java.lang.RuntimeException: java.util.regex.PatternSyntaxException: Unclosed counted closure near index 27
(?<pin>[A-Z][A-Z][A-Z0-9]{9)\Q[0-9]}\E$
                           ^
	at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:145)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:191)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartCallback(IsolatedDevModeMain.java:175)
	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:255)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:62)
	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$2.handle(VertxHttpHotReplacementSetup.java:52)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.util.regex.PatternSyntaxException: Unclosed counted closure near index 27
(?<pin>[A-Z][A-Z][A-Z0-9]{9)\Q[0-9]}\E$
                           ^
	at java.base/java.util.regex.Pattern.error(Pattern.java:2029)
	at java.base/java.util.regex.Pattern.closure(Pattern.java:3314)
	at java.base/java.util.regex.Pattern.sequence(Pattern.java:2215)
	at java.base/java.util.regex.Pattern.expr(Pattern.java:2070)
	at java.base/java.util.regex.Pattern.group0(Pattern.java:3006)
	at java.base/java.util.regex.Pattern.sequence(Pattern.java:2125)
	at java.base/java.util.regex.Pattern.expr(Pattern.java:2070)
	at java.base/java.util.regex.Pattern.compile(Pattern.java:1784)
	at java.base/java.util.regex.Pattern.<init>(Pattern.java:1431)
	at java.base/java.util.regex.Pattern.compile(Pattern.java:1070)
	at org.jboss.resteasy.reactive.server.mapping.URITemplate.<init>(URITemplate.java:154)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.buildResourceMethod(RuntimeResourceDeployment.java:120)
	at org.jboss.resteasy.reactive.server.core.startup.RuntimeDeploymentManager.deploy(RuntimeDeploymentManager.java:108)
	at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder.createDeployment(ResteasyReactiveRecorder.java:112)
	at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupEndpoints-1591811826.deploy_1(ResteasyReactiveProcessor$setupEndpoints-1591811826.zig:3800)
	at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupEndpoints-1591811826.deploy(ResteasyReactiveProcessor$setupEndpoints-1591811826.zig:45)
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:235)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:124)
	at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:346)
	at java.base/java.lang.Class.newInstance(Class.java:604)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:65)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:119)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at io.quarkus.runner.bootstrap.StartupActionImpl$3.run(StartupActionImpl.java:134)
	... 1 more
```


## To Reproduce
create project with quarkus-resteasy-reactive and resource:
```java
@Path("")
@Authenticated
public class Api {
    @GET
    @Path("/{pin: [A-Z][A-Z][A-Z0-9]{9}[0-9]}")
    public Uni<String> hello1(){
        return Uni.createFrom().optional(Optional.of("Hello, World 1!"));
    }

    @GET
    public Uni<String> hello2(){
        return Uni.createFrom().optional(Optional.of("Hello, World 2!"));
    }
}```

1. Create project
2. curl localhost:8080

openjdk version "14.0.1" 2020-04-14
tested with 1.13.0.Final and 1.12.0.Final
extensions: quarkus-resteasy-reactive, quarkus-elytron-security-properties-file

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/RequestMapper.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/RequestMapper.java
@@ -67,7 +67,7 @@ public class RequestMapper<T> {
                 if (segment.type == URITemplate.Type.CUSTOM_REGEX) {
                     Matcher matcher = segment.pattern.matcher(path);
                     matched = matcher.find(matchPos);
-                    if (!matched) {
+                    if (!matched || matcher.start() != matchPos) {
                         break;
                     }
                     matchPos = matcher.end();

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/URITemplate.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/URITemplate.java
@@ -40,6 +40,7 @@ public class URITemplate implements Dumpable, Comparable<URITemplate> {
         int litChars = 0;
         int capGroups = 0;
         int complexGroups = 0;
+        int bracesCount = 0;
         StringBuilder sb = new StringBuilder();
         int state = 0; //0 = start, 1 = parsing name, 2 = parsing regex
         for (int i = 0; i < template.length(); ++i) {
@@ -85,7 +86,7 @@ public class URITemplate implements Dumpable, Comparable<URITemplate> {
                     }
                     break;
                 case 2:
-                    if (c == '}') {
+                    if (c == '}' && bracesCount == 0) {
                         state = 0;
                         if (sb.length() > 0) {
                             capGroups++;
@@ -99,6 +100,11 @@ public class URITemplate implements Dumpable, Comparable<URITemplate> {
                         sb.setLength(0);
                     } else {
                         sb.append(c);
+                        if (c == '{') {
+                            bracesCount++;
+                        } else if (c == '}') {
+                            bracesCount--;
+                        }
                     }
                     break;
             }
@@ -116,6 +122,9 @@ public class URITemplate implements Dumpable, Comparable<URITemplate> {
             case 1:
             case 2:
                 throw new IllegalArgumentException("Invalid template " + template);
+        }
+        if (bracesCount > 0) {
+            throw new IllegalArgumentException("Invalid template " + template + " Unmatched { braces");
         }
 
         //coalesce the components

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/framework/ResteasyReactiveUnitTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/framework/ResteasyReactiveUnitTest.java
@@ -72,7 +72,7 @@ public class ResteasyReactiveUnitTest implements BeforeAllCallback, AfterAllCall
 
     static {
         System.setProperty("java.util.logging.manager", "org.jboss.logmanager.LogManager");
-        rootLogger = (Logger) LogManager.getLogManager().getLogger("");
+        rootLogger = LogManager.getLogManager().getLogger("");
     }
 
     private Path deploymentDir;
@@ -196,7 +196,7 @@ public class ResteasyReactiveUnitTest implements BeforeAllCallback, AfterAllCall
             vertx = Vertx.vertx();
             HttpServer server = vertx.createHttpServer();
             router = Router.router(vertx);
-            server.requestHandler(router).listen(8080);
+            server.requestHandler(router).listen(8080).toCompletionStage().toCompletableFuture().get();
             store.put(ResteasyReactiveUnitTest.class.getName(), new ExtensionContext.Store.CloseableResource() {
                 @Override
                 public void close() throws Throwable {
@@ -307,6 +307,7 @@ public class ResteasyReactiveUnitTest implements BeforeAllCallback, AfterAllCall
         if (deploymentDir != null) {
             deleteDirectory(deploymentDir);
         }
+        router.clear();
 
     }
 

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/RegexMatchTest.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/RegexMatchTest.java
@@ -1,0 +1,36 @@
+package org.jboss.resteasy.reactive.server.vertx.test.matching;
+
+import static org.hamcrest.Matchers.equalTo;
+
+import io.restassured.RestAssured;
+import java.util.function.Supplier;
+import org.jboss.resteasy.reactive.server.vertx.test.framework.ResteasyReactiveUnitTest;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class RegexMatchTest {
+
+    @RegisterExtension
+    static ResteasyReactiveUnitTest test = new ResteasyReactiveUnitTest()
+            .setArchiveProducer(new Supplier<JavaArchive>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClass(RegexResource.class);
+                }
+            });
+
+    @Test
+    public void testRegexMatch() {
+        RestAssured.get("/regex/1234")
+                .then()
+                .statusCode(200)
+                .body(equalTo("pin 1234"));
+        RestAssured.get("/regex/12345")
+                .then()
+                .statusCode(404);
+    }
+
+}

--- a/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/RegexResource.java
+++ b/independent-projects/resteasy-reactive/server/vertx/src/test/java/org/jboss/resteasy/reactive/server/vertx/test/matching/RegexResource.java
@@ -1,0 +1,16 @@
+package org.jboss.resteasy.reactive.server.vertx.test.matching;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+@Path("/regex")
+public class RegexResource {
+
+    @GET
+    @Path("/{pin:[A-Z0-9]{4}}")
+    public String hello(@PathParam("pin") String name) {
+        return "pin " + name;
+    }
+
+}

